File: backend/app/api/timeline.py
"""时间线与导出接口（Step 8）。

这一整组不依赖 ComfyUI / LLM。导出用原始素材，代理只服务预览。
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter
from pydantic import BaseModel, Field

from app.services.timeline import timeline

router = APIRouter(tags=["timeline"])


class AssembleBody(BaseModel):
    #: 默认 false = 对位调和（只更新自己铺过的片段，手工编辑活下来）。
    #: true 是兼容入口「彻底重铺」：先删掉所有 `origin="assembled"` 的片段再铺一遍，
    #: `manual` 的（用户自己加的、拆出去的声音）两种模式下都不碰。
    replace: bool = Field(
        default=False, description="true 先删掉装配铺的片段再重铺；false 对位调和"
    )


class MoveBody(BaseModel):
    start: float


class TrimBody(BaseModel):
    in_point: float | None = None
    out_point: float | None = None
    #: 拖左边缘时和 in_point 一起给：一次请求改完「从源素材哪里开始」与
    #: 「在时间线上哪里开始」，撤销栈上也只占一格（见 timeline.trim_clip）。
    start: float | None = None
    ripple: bool = Field(default=True, description="裁切后是否自动贴紧后续片段")


class SplitBody(BaseModel):
    at: float = Field(description="时间线上的绝对秒数")


class AudioSelectionBody(BaseModel):
    in_point: float
    out_point: float


class ReplaceVersionBody(BaseModel):
    version_id: str


class TrackBody(BaseModel):
    kind: str = Field(default="audio", description="video / audio / subtitle")
    name: str | None = Field(default=None, description="留空则自动编号（A1 占了就叫 A2）")


class TrackPatchBody(BaseModel):
    name: str | None = None
    muted: bool | None = None
    locked: bool | None = None


class AddClipBody(BaseModel):
    asset_id: str
    start: float = 0.0
    duration: float | None = Field(default=None, description="留空则用资产时长或 ffprobe 探测")
    label: str | None = None


class BlankClipBody(BaseModel):
    duration: float = Field(default=1.0, gt=0, description="黑场时长，单位秒")
    label: str | None = None


class BlankDurationBody(BaseModel):
    duration: float = Field(gt=0, description="黑场时长，单位秒")


class MoveTrackBody(BaseModel):
    track_id: str


class MixBody(BaseModel):
    muted: bool | None = None
    volume: float | None = Field(default=None, description="0 ~ 4，1 是原样")


class TransitionBody(BaseModel):
    from_clip_id: str
    to_clip_id: str
    kind: str = "dissolve"
    duration: float = 0.5


class ExportBody(BaseModel):
    path: str | None = Field(default=None, description="留空则写入工程 generations/exports/")


@router.get("/projects/{pid}/timeline")
async def get_timeline(pid: str) -> dict[str, Any]:
    return await timeline.get(pid)


@router.post("/projects/{pid}/timeline/assemble/plan")
async def assemble_plan(pid: str) -> dict[str, Any]:
    """只读账单：这次装配会动哪些片段。「重新装配」按下去之前先给人看一眼。"""
    return await timeline.assemble_plan(pid)


@router.post("/projects/{pid}/timeline/assemble")
async def auto_assemble(pid: str, body: AssembleBody) -> dict[str, Any]:
    return await timeline.auto_assemble(pid, replace=body.replace)


@router.post("/projects/{pid}/timeline/undo")
async def undo(pid: str) -> dict[str, Any]:
    return await timeline.undo(pid)


@router.post("/projects/{pid}/timeline/redo")
async def redo(pid: str) -> dict[str, Any]:
    return await timeline.redo(pid)


@router.post("/projects/{pid}/clips/{clip_id}/move")
async def move_clip(pid: str, clip_id: str, body: MoveBody) -> dict[str, Any]:
    return await timeline.move_clip(pid, clip_id, body.start)


@router.post("/projects/{pid}/clips/{clip_id}/trim")
async def trim_clip(pid: str, clip_id: str, body: TrimBody) -> dict[str, Any]:
    return await timeline.trim_clip(
        pid,
        clip_id,
        in_point=body.in_point,
        out_point=body.out_point,
        start=body.start,
        ripple=body.ripple,
    )


@router.post("/projects/{pid}/clips/{clip_id}/split")
async def split_clip(pid: str, clip_id: str, body: SplitBody) -> dict[str, Any]:
    return await timeline.split_clip(pid, clip_id, body.at)


@router.post("/projects/{pid}/clips/{clip_id}/isolate-audio-selection")
async def isolate_audio_selection(
    pid: str, clip_id: str, body: AudioSelectionBody
) -> dict[str, Any]:
    return await timeline.isolate_audio_selection(
        pid, clip_id, in_point=body.in_point, out_point=body.out_point
    )


@router.post("/projects/{pid}/clips/{clip_id}/clear")
async def clear_clip(pid: str, clip_id: str) -> dict[str, Any]:
    """清空这一段的内容（位置与长度留着，变成黑场 / 静音占位）。"""
    return await timeline.clear_clip(pid, clip_id)


@router.delete("/projects/{pid}/clips/{clip_id}")
async def delete_clip(pid: str, clip_id: str, ripple: bool = True) -> dict[str, Any]:
    return await timeline.delete_clip(pid, clip_id, ripple=ripple)


@router.post("/projects/{pid}/clips/{clip_id}/version")
async def replace_version(pid: str, clip_id: str, body: ReplaceVersionBody) -> dict[str, Any]:
    """只替换这一个片段的版本，整条时间线不重排。"""
    return await timeline.replace_version(pid, clip_id, body.version_id)


@router.post("/projects/{pid}/clips/{clip_id}/mix")
async def set_mix(pid: str, clip_id: str, body: MixBody) -> dict[str, Any]:
    """静音 / 音量。视频片段与音频片段用同一个入口。"""
    return await timeline.set_mix(pid, clip_id, muted=body.muted, volume=body.volume)


@router.post("/projects/{pid}/clips/{clip_id}/detach-audio", status_code=201)
async def detach_audio(pid: str, clip_id: str) -> dict[str, Any]:
    """把这段画面的声音拆成音频轨上的独立片段（源片段随之静音）。"""
    return await timeline.detach_audio(pid, clip_id)


@router.post("/projects/{pid}/clips/{clip_id}/blank-duration")
async def resize_blank_clip(pid: str, clip_id: str, body: BlankDurationBody) -> dict[str, Any]:
    return await timeline.resize_blank_clip(pid, clip_id, duration=body.duration)


@router.post("/projects/{pid}/clips/{clip_id}/audio-track")
async def move_clip_to_audio_track(pid: str, clip_id: str, body: MoveTrackBody) -> dict[str, Any]:
    return await timeline.move_clip_to_track(pid, clip_id, body.track_id)


@router.post("/projects/{pid}/clips/{clip_id}/new-audio-track", status_code=201)
async def move_clip_to_new_audio_track(pid: str, clip_id: str) -> dict[str, Any]:
    return await timeline.move_clip_to_new_audio_track(pid, clip_id)


@router.post("/projects/{pid}/tracks", status_code=201)
async def add_track(pid: str, body: TrackBody) -> dict[str, Any]:
    return await timeline.add_track(pid, kind=body.kind, name=body.name)


@router.patch("/projects/{pid}/tracks/{track_id}")
async def update_track(pid: str, track_id: str, body: TrackPatchBody) -> dict[str, Any]:
    return await timeline.update_track(
        pid, track_id, name=body.name, muted=body.muted, locked=body.locked
    )


@router.delete("/projects/{pid}/tracks/{track_id}")
async def delete_track(pid: str, track_id: str, force: bool = False) -> dict[str, Any]:
    """删轨道。上面还有片段时先回 CONFLICT + `confirm: "force"`，确认后带 `?force=true` 重放。"""
    return await timeline.delete_track(pid, track_id, force=force)


@router.post("/projects/{pid}/tracks/{track_id}/clips", status_code=201)
async def add_clip(pid: str, track_id: str, body: AddClipBody) -> dict[str, Any]:
    """把一个已登记的资产放到轨道上（导入的配乐 / 音效走这里）。"""
    return await timeline.add_clip(
        pid,
        track_id,
        body.asset_id,
        start=body.start,
        duration=body.duration,
        label=body.label,
    )


@router.post("/projects/{pid}/tracks/{track_id}/blank-clips", status_code=201)
async def add_blank_clip(pid: str, track_id: str, body: BlankClipBody) -> dict[str, Any]:
    return await timeline.add_blank_clip(pid, track_id, duration=body.duration, label=body.label)


@router.get("/projects/{pid}/transitions")
async def list_transitions(pid: str) -> list[dict[str, Any]]:
    return await timeline.list_transitions(pid)


@router.post("/projects/{pid}/transitions", status_code=201)
async def add_transition(pid: str, body: TransitionBody) -> dict[str, Any]:
    return await timeline.add_transition(
        pid, body.from_clip_id, body.to_clip_id, body.kind, body.duration
    )


@router.delete("/projects/{pid}/transitions/{tid}", status_code=204)
async def delete_transition(pid: str, tid: str) -> None:
    await timeline.delete_transition(pid, tid)


@router.post("/projects/{pid}/assets/{asset_id}/proxy")
async def ensure_proxy(pid: str, asset_id: str) -> dict[str, Any]:
    return await timeline.ensure_proxy(pid, asset_id)


@router.get("/projects/{pid}/exports")
async def list_exports(pid: str) -> list[dict[str, Any]]:
    return await timeline.list_exports(pid)


@router.post("/projects/{pid}/exports/open-folder")
async def open_export_folder(pid: str) -> dict[str, str]:
    return await timeline.open_export_folder(pid)


@router.get("/projects/{pid}/export/command")
async def export_command(pid: str) -> dict[str, Any]:
    """导出前的预检：把将要执行的 FFmpeg 命令原样给人看。

    `warnings` 里是「会被丢掉 / 说不准」的那些事（静音的音频轨、比画面长的声音、
    ffprobe 不在所以不知道某段画面有没有声音）——不拦导出，但必须显示出来。
    """
    plan = await timeline.build_command(pid)
    return {
        "path": plan["path"],
        "command": plan["command"],
        "clips": len(plan["clips"]),
        "audio_clips": len(plan["audio_clips"]),
        "warnings": plan["warnings"],
    }


@router.post("/projects/{pid}/export", status_code=201)
async def export(pid: str, body: ExportBody) -> dict[str, Any]:
    return await timeline.export(pid, body.path)

File: frontend/src/stores/timeline.ts
/**
 * 时间线状态（Step 8 的前端）。
 *
 * 三个刻意的取舍：
 *   1. **每条编辑命令的返回值就是新的整条时间线**，直接整体覆盖。后端已经重排过
 *      index_no 与 ripple 之后的位置，前端再算一遍只会算出第二套真相。
 *   2. **撤销可用性只看后端**（`can_undo` / `can_redo`）。撤销栈在进程里，重启即清空——
 *      前端自己记一份必然会在重启后骗人。
 *   3. **导出分两步**：先 `plan()` 把将要执行的 FFmpeg 命令摆出来，再 `runExport()`。
 *      这里不把两步合成一个「导出」动作，因为看不见命令的导出出问题时无从下手。
 *
 * 装配的 `skipped` 单独存一份：它是这一页最重要的产出之一（「哪几个镜头没铺上、为什么」），
 * 不能只作为一次性返回值被丢掉。
 */

import { defineStore } from 'pinia'
import { computed, ref } from 'vue'
import { ApiError } from '@/shared/api/client'
import {
  timelineApi,
  type AssembleResult,
  type Clip,
  type DetachResult,
  type ExportPlan,
  type ExportRecord,
  type Timeline,
  type Track,
  type Transition,
  type TrimBody,
} from '@/shared/api/timeline'

export const useTimelineStore = defineStore('timeline', () => {
  const timeline = ref<Timeline | null>(null)
  const transitions = ref<Transition[]>([])
  const exports = ref<ExportRecord[]>([])
  /** 最近一次装配被跳过的镜头，带结构化理由。 */
  const skipped = ref<AssembleResult['skipped']>([])
  const placedCount = ref<number | null>(null)
  /** 导出预检结果；`null` 表示还没预检过。 */
  const plan = ref<ExportPlan | null>(null)
  /** 最近一次「拆出声音」的结果，界面要据此说明新开了轨 / 复用了文件。 */
  const lastDetach = ref<DetachResult | null>(null)

  const busy = ref(false)
  const lastError = ref<ApiError | null>(null)

  const clips = computed<Clip[]>(() => timeline.value?.tracks.flatMap((t) => t.clips) ?? [])
  /** 文件已经不在磁盘上的片段——导出前必须先处理，否则 FFmpeg 会在半路失败。 */
  const missing = computed(() => clips.value.filter((c) => c.missing_file))
  const canUndo = computed(() => timeline.value?.can_undo ?? false)
  const canRedo = computed(() => timeline.value?.can_redo ?? false)
  /**
   * 画面来自**第一条视频轨**——和导出（`timeline.build_command`）同一条规矩。
   * 预览器要是自己另挑一条，看到的就不是将要导出的东西。
   */
  const videoTrack = computed<Track | null>(
    () => timeline.value?.tracks.find((t) => t.kind === 'video') ?? null,
  )
  const audioTracks = computed<Track[]>(
    () => timeline.value?.tracks.filter((t) => t.kind === 'audio') ?? [],
  )

  function clearError(): void {
    lastError.value = null
  }

  function clearAssembleNote(): void {
    skipped.value = []
    placedCount.value = null
  }

  async function guarded<T>(run: () => Promise<T>): Promise<T> {
    busy.value = true
    try {
      const out = await run()
      lastError.value = null
      return out
    } catch (err) {
      lastError.value = err instanceof ApiError ? err : null
      throw err
    } finally {
      busy.value = false
    }
  }

  /** 转场与导出历史各自独立成败：其中一个拉不到不该把时间线清空。 */
  async function load(pid: string): Promise<void> {
    await guarded(async () => {
      timeline.value = await timelineApi.get(pid)
    })
    const [tr, ex] = await Promise.all([
      timelineApi.transitions(pid).catch(() => [] as Transition[]),
      timelineApi.exports(pid).catch(() => [] as ExportRecord[]),
    ])
    transitions.value = tr
    exports.value = ex
  }

  async function assemble(pid: string, replace = true): Promise<void> {
    await guarded(async () => {
      const out = await timelineApi.assemble(pid, replace)
      timeline.value = out.timeline
      skipped.value = out.skipped
      placedCount.value = out.placed.length
    })
  }

  /** 撤销栈为空时后端回 CONFLICT——照常显示，不吞。 */
  async function undo(pid: string): Promise<void> {
    await guarded(async () => {
      timeline.value = await timelineApi.undo(pid)
    })
  }

  async function redo(pid: string): Promise<void> {
    await guarded(async () => {
      timeline.value = await timelineApi.redo(pid)
    })
  }

  async function move(pid: string, clipId: string, start: number): Promise<void> {
    await guarded(async () => {
      timeline.value = await timelineApi.move(pid, clipId, start)
    })
  }

  /**
   * 裁切。**拖左边缘时 `in_point` 与 `start` 在同一个 body 里**：一次请求、一格撤销——
   * 分两次发的话「撤销」只能退回一半，边缘会停在一个用户从来没见过的位置。
   */
  async function trim(pid: string, clipId: string, body: TrimBody): Promise<void> {
    await guarded(async () => {
      timeline.value = await timelineApi.trim(pid, clipId, body)
    })
  }

  async function split(pid: string, clipId: string, at: number): Promise<void> {
    await guarded(async () => {
      timeline.value = await timelineApi.split(pid, clipId, at)
    })
  }

  async function isolateAudioSelection(
    pid: string,
    clipId: string,
    body: { in_point: number; out_point: number },
  ): Promise<{ selectedClipId: string; segments: number }> {
    const out = await guarded(() => timelineApi.isolateAudioSelection(pid, clipId, body))
    timeline.value = out.timeline
    return { selectedClipId: out.selected_clip_id, segments: out.segments }
  }

  async function remove(pid: string, clipId: string, ripple = true): Promise<void> {
    await guarded(async () => {
      timeline.value = await timelineApi.remove(pid, clipId, ripple)
    })
    // 删片段会连带删掉挂在它上面的转场
    transitions.value = await timelineApi.transitions(pid).catch(() => transitions.value)
  }

  /** 清空这一段的内容（位置与长度不动）。删除是另一件事，见 `remove`。 */
  async function clear(pid: string, clipId: string): Promise<void> {
    await guarded(async () => {
      timeline.value = await timelineApi.clear(pid, clipId)
    })
  }

  async function replaceVersion(pid: string, clipId: string, versionId: string): Promise<void> {
    await guarded(async () => {
      timeline.value = await timelineApi.replaceVersion(pid, clipId, versionId)
    })
  }

  // --- 声音 ---

  async function setMix(
    pid: string,
    clipId: string,
    body: { muted?: boolean; volume?: number },
  ): Promise<void> {
    await guarded(async () => {
      timeline.value = await timelineApi.setMix(pid, clipId, body)
    })
  }

  /**
   * 把画面的声音拆成独立音频片段。返回值留一份给界面说话：
   * 新开了一条轨、或者复用了上次拆好的文件，都是用户该知道的事。
   */
  async function detachAudio(pid: string, clipId: string): Promise<DetachResult> {
    const out = await guarded(() => timelineApi.detachAudio(pid, clipId))
    timeline.value = out.timeline
    lastDetach.value = out
    return out
  }

  // --- 轨道 ---

  async function addTrack(pid: string, kind = 'audio', name?: string | null): Promise<Track> {
    const out = await guarded(() => timelineApi.addTrack(pid, kind, name))
    timeline.value = out.timeline
    return out.track
  }

  async function patchTrack(
    pid: string,
    trackId: string,
    body: { name?: string; muted?: boolean; locked?: boolean },
  ): Promise<void> {
    await guarded(async () => {
      timeline.value = await timelineApi.patchTrack(pid, trackId, body)
    })
  }

  /** 轨道上还有片段时后端先回 CONFLICT + `confirm: "force"`，界面确认后再带 `force` 重放。 */
  async function removeTrack(pid: string, trackId: string, force = false): Promise<void> {
    await guarded(async () => {
      timeline.value = await timelineApi.removeTrack(pid, trackId, force)
    })
  }

  async function addClip(
    pid: string,
    trackId: string,
    body: { asset_id: string; start?: number; duration?: number | null; label?: string | null },
  ): Promise<string> {
    const out = await guarded(() => timelineApi.addClip(pid, trackId, body))
    timeline.value = out.timeline
    return out.clip_id
  }

  async function addBlankClip(
    pid: string,
    trackId: string,
    body: { duration: number; label?: string | null },
  ): Promise<string> {
    const out = await guarded(() => timelineApi.addBlankClip(pid, trackId, body))
    timeline.value = out.timeline
    return out.clip_id
  }

  async function resizeBlankClip(pid: string, clipId: string, duration: number): Promise<void> {
    await guarded(async () => {
      timeline.value = await timelineApi.resizeBlankClip(pid, clipId, duration)
    })
  }

  async function moveToAudioTrack(pid: string, clipId: string, trackId: string): Promise<void> {
    await guarded(async () => {
      timeline.value = await timelineApi.moveToAudioTrack(pid, clipId, trackId)
    })
  }

  async function moveToNewAudioTrack(pid: string, clipId: string): Promise<string | null> {
    const out = await guarded(() => timelineApi.moveToNewAudioTrack(pid, clipId))
    timeline.value = out.timeline
    return out.track_id
  }

  async function addTransition(
    pid: string,
    body: { from_clip_id: string; to_clip_id: string; kind?: string; duration?: number },
  ): Promise<void> {
    await guarded(async () => {
      await timelineApi.addTransition(pid, body)
      transitions.value = await timelineApi.transitions(pid)
    })
  }

  async function removeTransition(pid: string, tid: string): Promise<void> {
    await guarded(async () => {
      await timelineApi.removeTransition(pid, tid)
      transitions.value = await timelineApi.transitions(pid)
    })
  }

  /** 预检：只拿命令，不起进程。 */
  async function loadPlan(pid: string): Promise<void> {
    await guarded(async () => {
      plan.value = await timelineApi.exportCommand(pid)
    })
  }

  async function runExport(pid: string, path?: string | null): Promise<ExportRecord | null> {
    let record: ExportRecord | null = null
    try {
      record = await guarded(() => timelineApi.export(pid, path))
    } catch {
      record = null
    }
    // 成功与失败都会留一条记录，历史必须刷新
    exports.value = await timelineApi.exports(pid).catch(() => exports.value)
    return record
  }

  async function openExportFolder(pid: string): Promise<string | null> {
    try {
      return (await guarded(() => timelineApi.openExportFolder(pid))).path
    } catch {
      return null
    }
  }

  return {
    timeline,
    transitions,
    exports,
    skipped,
    placedCount,
    plan,
    lastDetach,
    busy,
    lastError,
    clips,
    missing,
    canUndo,
    canRedo,
    videoTrack,
    audioTracks,
    load,
    assemble,
    undo,
    redo,
    move,
    trim,
    split,
    isolateAudioSelection,
    remove,
    clear,
    replaceVersion,
    setMix,
    detachAudio,
    addTrack,
    patchTrack,
    removeTrack,
    addClip,
    addBlankClip,
    resizeBlankClip,
    moveToAudioTrack,
    moveToNewAudioTrack,
    addTransition,
    removeTransition,
    loadPlan,
    runExport,
    openExportFolder,
    clearError,
    clearAssembleNote,
  }
})
